mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -0,0 +1,652 @@
+//===- MemoryAccessAnalysis.cpp - SYCL Memory Access Analysis -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h"
+#include "mlir/Analysis/DataFlow/IntegerRangeAnalysis.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Debug.h"
+#include <utility>
+
+#define DEBUG_TYPE "memory-access-analysis"
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Determine whether a value is an known integer value.
+static Optional<APInt> getConstIntegerValue(Value val, DataFlowSolver &solver) {
+  if (!isa<IntegerType>(val.getType()))
```

Comment:
does this allow index type? how about 
```suggestion
  if (!val.getType().isIntOrIndex())
```